crates/secrets-nats-kv/src/api.rs

```diff
@@ -0,0 +1,660 @@
+                        .await;
+                }
+            }
+        }
+
+        Ok(())
+    }
+
+    async fn get_lock(&self, subject: String) -> anyhow::Result<PublishAck> {
+        // TODO: make this all configurable
+        let exp = ExponentialBackoffBuilder::new()
+            .with_initial_interval(Duration::from_millis(100))
+            .with_max_elapsed_time(Some(Duration::from_secs(3)))
+            .build();
+
+        let op = || async {
+            let resp = self
+                .client
+                .request(subject.clone(), "lock".into())
+                .await
+                .map_err(|e| {
+                    println!("Error locking state stream: {}", e);
+                    error!("Error locking state stream: {}", e);
+                    BackoffError::transient("")
+                })?;
+            match serde_json::from_slice(&resp.payload) {
+                Ok(Response::Ok(p)) => Ok(p),
+                Ok(Response::Err { error: e }) => {
+                    println!("Error locking state stream: {:?}", e);
+                    error!("Error locking state stream: {:?}", e);
+                    Err(BackoffError::transient("unable to get lock"))
+                }
+                Err(e) => {
+                    println!("Error locking state stream: {}", e);
+                    error!("Error locking state stream: {}", e);
+                    Err(BackoffError::permanent("error publishing message"))
+                }
+            }
+        };
+
+        let result = retry(exp, op).await;
+        result.map_err(|_e| anyhow::anyhow!("timed out getting lock"))
+    }
+
+    // TODO: add a way to specify labels that should apply to this mapping. That way you can
+    // provide host labels that should grant an entity access to a secret.
+    async fn add_mapping(&self, entity: String, values: HashSet<String>) -> anyhow::Result<()> {
+        let c = jetstream::new(self.client.clone());
+        let subject = format!("{}.{}", self.lock_stream_name(), entity);
+
+        let ack = self.get_lock(subject.clone()).await?;
+        let seq = ack.sequence;
+        let state = self.state_bucket().await?;
+        let entry = state.get(&entity).await?;
+        if let Some(e) = entry {
+            let mut stored_values: HashSet<String> = serde_json::from_slice(&e)?;
+            stored_values.extend(values.clone());
+            let str = serde_json::to_string(&stored_values)?;
+            state.put(entity.clone(), str.into()).await?;
+        } else {
+            let str = serde_json::to_string(&values)?;
+            state.put(entity.clone(), str.into()).await?;
+        }
+        let s = c.get_stream(&self.lock_stream_name()).await?;
+        s.delete_message(seq).await?;
+        Ok(())
+    }
+
+    async fn remove_mapping(&self, entity: String, values: HashSet<String>) -> anyhow::Result<()> {
+        let c = jetstream::new(self.client.clone());
```

Comment:
Nit: The logic around getting the lock and the stream could probably be abstracted out of this function and `add_mapping` to reduce duplication. Probably could implement the guard thing you mention in the TODO below